src/ert/parsing/config_keywords.py

```diff
@@ -167,55 +176,83 @@ class SchemaItem(BaseModel):
     # Allowed values for specific arguments, if no entry, all values allowed
     indexed_selection_set: Mapping[int, List[str]] = {}
 
-    def check_valid(self, val: "FileContextToken", index: int) -> None:
-        if (
+    def _is_in_allowed_values_for_kw(
+        self, token: "FileContextToken", index: int
+    ) -> None:
+        return not (
             index in self.indexed_selection_set
-            and val not in self.indexed_selection_set[index]
-        ):
-            raise ConfigValidationError(
-                f"{self.kw!r} argument {index!r} must be one of"
-                f" {self.indexed_selection_set[index]!r} was {val.value!r}",
-                config_file=val.filename,
+            and token not in self.indexed_selection_set[index]
+        )
+
+    def token_to_value_with_context(
+        self, token: FileContextToken, index: int, keyword: FileContextToken
+    ) -> Optional[PrimitiveWithContext]:
+        """
+        Converts a FileContextToken to a typed primitive that
+        behaves like a primitive, but also contains its location in the file,
+        as well the keyword it pertains to and its location in the file.
+
+        :param token: the token to be converted
+        :param index: the index of the token
+        :param keyword: the keyword it pertains to
+
+        :return: The token as a primitive with context of itself and its keyword
+        """
+        # pylint: disable=too-many-return-statements, too-many-branches
+
+        if not self._is_in_allowed_values_for_kw(token, index):
+            raise ConfigValidationError.from_info(
+                ErrorInfo(
+                    message=f"{self.kw!r} argument {index!r} must be one of"
+                    f" {self.indexed_selection_set[index]!r} was {token.value!r}",
+                    filename=token.filename,
+                    originates_from=token,
+                )
             )
 
-    def convert(
-        self, token: FileContextToken, index: int
-    ) -> Optional[Union[str, int, float]]:
-        # pylint: disable=too-many-branches, too-many-return-statements
-        self.check_valid(token, index)
         if not len(self.type_map) > index:
-            return token
+            return StringToken(str(token), token, keyword)
```

Comment:
Mypy is complaining that `StringToken` is not a `Optional[PrimitiveWithContext]`

Reply:
Fixing by setting `PrimitiveWithContext` to a union of all the "typed tokens"